client.ts

```diff
@@ -1,44 +1,67 @@
 import { dial } from "deno";
 import { Connection } from "./connection.ts";
+import { Pool } from "./pool.ts";
 import { Query, QueryConfig, QueryResult } from "./query.ts";
-import { IConnectionParams, ConnectionParams } from "./connection_params.ts";
+import { ConnectionParams, IConnectionParams } from "./connection_params.ts";
 
 export class Client {
-  connection: Connection;
-  connectionParams: ConnectionParams;
+  private _connection: Connection;
+  private _connectionParams: ConnectionParams;
+  release: () => void;
 
-  constructor(config?: IConnectionParams | string) {
-    this.connectionParams = new ConnectionParams(config);
+  constructor(
+    config?: IConnectionParams | string | Connection,
+    release?: () => void
+  ) {
+    if (config instanceof Connection) {
+      this._connection = config;
+      this.release = function() {
+        release();
+        delete this._connection;
+        delete this.release;
+      };
+    } else {
+      this._connectionParams = new ConnectionParams(config);
+    }
   }
 
-  async connect() {
-    const { host, port } = this.connectionParams;
+  private async newConnection(connectionParams: ConnectionParams) {
+    const { host, port } = connectionParams;
     let addr = `${host}:${port}`;
 
     const conn = await dial("tcp", addr);
-    this.connection = new Connection(conn, this.connectionParams);
+    const connection = new Connection(conn, connectionParams);
 
-    await this.connection.startup({ ...this.connectionParams });
+    await connection.startup({ ...connectionParams });
+    await connection.initSQL();
```

Comment:
Is it needed? I feel like executing query automatically every time we connect is not a good thing.

Reply:
This is pretty standard in creating a pool in other systems.

Reply:
Could you link to some examples? I feel it's superfluous to add this query, but I'm not 100% sure.

Reply:
https://stackoverflow.com/q/3668506/1240268
This is in jdbc/spring/play.

python: https://docs.sqlalchemy.org/en/latest/core/pooling.html#disconnect-handling-pessimistic

Basically you want to test the connection for idleness, and verifying it's connected. `select 1` is very cheap for this purpose (no disk access). You are only doing this once per connection (or once during each idle period) to keep the connection alive.

Reply:
Okay sounds reasonable, could you add a comment about that?